python_ta/checkers/one_iteration_checker.py

```diff
@@ -25,20 +26,63 @@ class OneIterationChecker(BaseChecker):
     # pass in message symbol as a parameter of check_messages
     @check_messages('one-iteration')
     def visit_for(self, node):
-        if self._check_one_iteration(node):
+        if hasattr(node, 'cfg_block') and self._check_one_iteration(node):
             self.add_message('one-iteration', node=node)
 
     @check_messages('one-iteration')
     def visit_while(self, node):
-        if self._check_one_iteration(node):
+        if hasattr(node, 'cfg_block') and self._check_one_iteration(node):
             self.add_message('one-iteration', node=node)
 
     def _check_one_iteration(self, node: Union[astroid.For, astroid.While]) -> bool:
-        """Return whether the given loop is guaranteed to stop after one iteration."""
-        # TODO: Complete this method.
-        # print(node.cfg_block)
+        """Return whether the given loop is guaranteed to stop after one iteration.
 
-        return False
+        More precisely, Returns False if there exists a direct predecessor
+        block `p` to the start of the loop block `s` such that the
+        first statement in `p` is a child node of <node> and that there exists a
+        path from `s` to `p.
+
+        Note: For `while` loops, 'start of the loop block' refers to the block with
+        the test condition (or the first of the blocks that make up test condition).
+        For `for` loops, it refers to the block with the assignment target.
+        """
+        start = node.target if isinstance(node, astroid.For) else node
+        preds = start.cfg_block.predecessors
+
+        if preds == []:
+            return False
+
+        for pred in preds:
+            pred = pred.source
+            if node.parent_of(pred.statements[0]) and \
+                    self.is_predecessor(pred, start.cfg_block):
```

Comment:
I think I'm forgetting something here; why is the `self.is_predecessor` check necessary? I.e., why isn't it enough that a predecessor contains a statement that's inside the given loop?

Reply:
The reason behind this is because of how unreachable blocks (statements after `jump` statements) within the loops are handled. If there is a statement `s` after a jump statement then it will be in a separate block that 'points' to the start block of the loop. This occurs because of these two lines:
https://github.com/pyta-uoft/pyta/blob/a36e750b130c4ab23294f24273016e9e96ea2687/python_ta/cfg/visitor.py#L201-L202

This means that a predecessor to the loop start block could be one of the unreachable blocks that won't be executed.